npm/pkg/dataplane/ipsets/ipset.go

```diff
@@ -0,0 +1,191 @@
+package ipsets
+
+import (
+	"fmt"
+
+	"github.com/Azure/azure-container-networking/npm/util"
+)
+
+type IPSet struct {
+	Name       string
+	HashedName string
+	Type       SetType
+	// IpPodKey is used for setMaps to store Ips and ports as keys
+	// and podKey as value
+	IPPodKey map[string]string
+	// This is used for listMaps to store child IP Sets
+	MemberIPSets map[string]*IPSet
+	// Using a map here to emulate set of netpol names
+	SelectorReference map[string]struct{}
+	NetPolReference   map[string]struct{}
+	IpsetReferCount   int
+}
+
+type SetType int32
+
+const (
+	Unknown                  SetType = 0
+	NameSpace                SetType = 1
+	KeyLabelOfNameSpace      SetType = 2
+	KeyValueLabelOfNameSpace SetType = 3
+	KeyLabelOfPod            SetType = 4
+	KeyValueLabelOfPod       SetType = 5
+	NamedPorts               SetType = 6
+	NestedLabelOfPod         SetType = 7
+	CIDRBlocks               SetType = 8
+)
+
+var SetTypeName = map[int32]string{
+	0: "Unknown",
+	1: "NameSpace",
+	2: "KeyLabelOfNameSpace",
+	3: "KeyValueLabelOfNameSpace",
+	4: "KeyLabelOfPod",
+	5: "KeyValueLabelOfPod",
+	6: "NamedPorts",
+	7: "NestedLabelOfPod",
+	8: "CIDRBlocks",
+}
+
+var SetTypeValue = map[string]int32{
+	"Unknown":                  0,
+	"NameSpace":                1,
+	"KeyLabelOfNameSpace":      2,
+	"KeyValueLabelOfNameSpace": 3,
+	"KeyLabelOfPod":            4,
+	"KeyValueLabelOfPod":       5,
+	"NamedPorts":               6,
+	"NestedLabelOfPod":         7,
+	"CIDRBlocks":               8,
+}
+
+func (x SetType) String() string {
+	return SetTypeName[int32(x)]
+}
+
+func GetSetType(x string) SetType {
+	return SetType(SetTypeValue[x])
+}
+
+type SetKind string
+
+const (
+	ListSet SetKind = "list"
+	HashSet SetKind = "set"
+)
+
+func NewIPSet(name string, setType SetType) *IPSet {
+	set := &IPSet{
+		Name:              name,
+		HashedName:        util.GetHashedName(name),
+		Type:              setType,
+		SelectorReference: make(map[string]struct{}),
```

Comment:
is the empty struct a placeholder? Or what does it mean?

Reply:
My intension was to emulate a "set" behavior, https://www.davidkaya.com/sets-in-golang/ .. so that we do not have to check for duplicates everytime.

Reply:
What do we think of encapsulating set behavior in a new struct like they suggest in that article? This may be natural to experienced GoLangers, but I think it would improve readability, perhaps be less bug prone, and wouldn't be wasted energy since we use the data structure in 4 different fields. Thoughts @rbtr @matmerr @JungukCho ?
```
type set struct {
    m map[string]struct{}
}

func NewSet() *set {
    s := &set{}
    s.m = make(map[string]struct{})
}

func (s *set) Add(value string) {
    s.m[value] = exists
}

func (s *set) Remove(value string) {
    delete(s.m, value)
}

func (s *set) Contains(value string) bool {
    _, c := s.m[value]
    return c
}
```

Reply:
I'm lukewarm on a Set type,  I don't think it would simplify the code in reality and is unusual enough in go that I would then have to go find the custom type definition to see how it had been implemented.
BUT, I'm very used to using maps for this, so 🤷🏼‍♂️

Reply:
i am okay with either approach, no major preference, but i felt the Map approach keeps it simple and less cognitive load for reviewer. I understand this question of why struct may arise again, i will leave a comment explaining the same in code.

Reply:
If you look at our vendor files apimachinery or grpc lib uses this same notion.

Reply:
like @huntergregory alluded to, this should make sense to gophers immediately, it's common to see. but it's not obvious the first time you see it - a comment saying "using a set" or something would help 👍🏼

Reply:
I don't think you repeat below comment whenever using set. 
```go
// Using a map to emulate set and value as struct{} for
// minimal memory consumption
```
Instead, for example, shortly. It will be better comments to improve readability. Especially, more important information is to specify what key and values are when map and set are used in the code.
```go
set using map, key: (describe what key is)
```